File: src/leetcode/python/SymmetricTree_101.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class SymmetricTree_101(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def symmetric(left, right):
            if not left and not right:
                return True
            elif left and not right or not left and right:
                return False
            if left.val != right.val:
                return False
            if not symmetric(left.left, right.right):
                return False
            return symmetric(left.right, right.left)

        if not root:
            return True
        return symmetric(root.left, root.right)

File: src/leetcode/java/SymmetricTree_101.java
package leetcode.java.easy;

public class SymmetricTree_101 {
  public boolean isSymmetric(TreeNode root) {
    if (root == null) return true;
    return symmetric(root.left, root.right);
  }

  public static boolean symmetric(TreeNode left, TreeNode right) {
    if (left == null && right == null) {
      return true;
    } else if ((left != null && right == null) || (left == null && right != null)) {
      return false;
    }
    if (left.val != right.val) return false;
    if (!symmetric(left.left, right.right)) return false;
    return symmetric(left.right, right.left);
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
  }
}
